### PR TITLE
refact: the DPA2 descriptor

### PR DESCRIPTION
**Pull request by iProzd**
- Refact the DPA2 descriptor in PyTorch with clearer interface
- Support residual
- Remove bn
- Add numpy implement

## Summary by CodeRabbit

- **New Features**
  - Added a new descriptor class `DescrptDPA2` implementing DPA-2 functionality for computing descriptors and representations based on input coordinates and atom types.
  - Expanded supported backends for DPA-2 descriptor to include DP in addition to PyTorch.

- **Documentation**
  - Updated the supported backends information in the documentation for the DPA-2 descriptor to reflect the addition of DP backend support.
  - Added a reference to the model implementation and a training example link in the DPA-2 descriptor documentation.

- **Tests**
  - Introduced test cases for the `DescrptDPA2` class in different frameworks like TensorFlow, PyTorch, and DeepMD to cover various parameters and configurations.
  - Validated the functionality of the `DescrptDPA2` descriptor class for deep learning models in the test case class `TestDescrptDPA2`.

### CODE REVIEW

**Review by njzjz — CHANGES_REQUESTED** [review]

File: deepmd/pt/model/descriptor/dpa2.py [hunk 1]
```diff
@@ -75,118 +91,171 @@ def __init__(
...
         repformer_attn2_nhead: int = 4,
         repformer_attn2_has_gate: bool = False,
-        repformer_activation: str = "tanh",
+        repformer_activation_function: str = "tanh",
         repformer_update_style: str = "res_avg",
-        repformer_set_davg_zero: bool = True,  # TODO
-        repformer_add_type_ebd_to_seq: bool = False,
+        repformer_update_residual: float = 0.001,
+        repformer_update_residual_init: str = "norm",
+        repformer_set_davg_zero: bool = True,
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/pt/model/descriptor/dpa2.py:98]: good idea.

File: deepmd/utils/argcheck.py [hunk 2]
```diff
@@ -2341,6 +2452,25 @@ def gen_args(**kwargs) -> List[Argument]:
     ]
 
 
+def backend_compat(data):
```
[the remark this answers is not in the document]
Comment (iProzd) [reply deepmd/utils/argcheck.py]: Yes, this is caused by a branch chosen mistake. I've updated.

File: deepmd/pt/model/descriptor/se_atten.py [hunk 3]
```diff
@@ -808,37 +808,200 @@ def deserialize(cls, data: dict) -> "NeighborGatedAttentionLayer":
         return obj
 
 
+# class GatedAttentionLayer(nn.Module):
```
[the remark this answers is not in the document]
Comment (iProzd) [reply deepmd/pt/model/descriptor/se_atten.py]: It's a backup for old impl and forget to remove. I've updated.

File: deepmd/pt/model/descriptor/se_atten.py [hunk 4]
```diff
@@ -808,37 +808,200 @@ def deserialize(cls, data: dict) -> "NeighborGatedAttentionLayer":
...
+        num_heads: int = 1,
         dotr: bool = False,
         do_mask: bool = False,
         scaling_factor: float = 1.0,
         normalize: bool = True,
         temperature: Optional[float] = None,
         bias: bool = True,
         smooth: bool = True,
-        precision: str = DEFAULT_PRECISION,
+        precision: str = "float",
```
[the remark this answers is not in the document]
Comment (iProzd) [reply deepmd/pt/model/descriptor/se_atten.py]: Yes, I've updated.

File: deepmd/pt/model/descriptor/repformer_layer.py [hunk 5]
```diff
@@ -1,125 +1,324 @@
...
+    # nlist: nf x nloc x nnei
+    nf, nloc, nnei = nlist.shape
+    # g1_ext: nf x nall x ng1
     ng1 = g1_ext.shape[-1]
-    # index: nb x (nloc x nnei) x ng1
-    index = nlist.reshape(nb, nloc * nnei).unsqueeze(-1).expand(-1, -1, ng1)
-    # gg1  : nb x (nloc x nnei) x ng1
+    # index: nf x (nloc x nnei) x ng1
+    index = nlist.reshape(nf, nloc * nnei).unsqueeze(-1).expand(-1, -1, ng1)
+    # gg1  : nf x (nloc x nnei) x ng1
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/descriptor/repformer_layer.py]: not recommended to change "number of batch" to "number of frames" they basically mean the same thing...

File: deepmd/pt/model/descriptor/repformer_layer.py [hunk 6]
```diff
@@ -431,165 +872,74 @@ def _update_g1_conv(
...
-        else:
-            g2 = _apply_switch(g2, sw)
-            invnnei = (1.0 / float(nnei)) * torch.ones(
-                (nb, nloc, 1, 1), dtype=env.GLOBAL_PT_FLOAT_PRECISION, device=g2.device
-            )
-        # nb x nloc x 3 x ng2
-        h2g2 = torch.matmul(torch.transpose(h2, -1, -2), g2) * invnnei
-        return h2g2
-
-    def _cal_grrg(self, h2g2: torch.Tensor) -> torch.Tensor:
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/descriptor/repformer_layer.py:483]: I know self is not used. but do not move the position of cal_grrg. you may make it a static method if you want.

File: deepmd/pt/model/descriptor/repformer_layer.py [hunk 7]
```diff
@@ -431,165 +872,74 @@ def _update_g1_conv(
...
             invnnei = (1.0 / float(nnei)) * torch.ones(
-                (nb, nloc, 1), dtype=env.GLOBAL_PT_FLOAT_PRECISION, device=gg1.device
+                (nf, nloc, 1), dtype=gg1.dtype, device=gg1.device
             )
-        # nb x nloc x ng2
+        # nf x nloc x ng2
         g1_11 = torch.sum(g2 * gg1, dim=2) * invnnei
         return g1_11
 
-    def _cal_h2g2(
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/descriptor/repformer_layer.py:455]: I know self is not used but do not move the position of cal_h2g2. you may make it a static method if you want

File: deepmd/pt/model/descriptor/repformer_layer.py [hunk 8]
```diff
@@ -1,125 +1,324 @@
...
+        and remains 0 beyond rcut, with shape nf x nloc x nnei.
+    """
+    # gg:  nf x nloc x nnei x d
     # sw:  nf x nloc x nnei
     return gg * sw.unsqueeze(-1)
 
 
-def _apply_h_norm(
-    hh: torch.Tensor,  # nf x nloc x nnei x 3
+def _cal_hg(
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/descriptor/repformer_layer.py:140]: why renaming cal_h2g2 to cal_hg? any reason?

File: deepmd/pt/model/descriptor/repformer_layer.py [hunk 9]
```diff
@@ -442,154 +898,48 @@ def _update_g1_conv(
...
-        nb, nloc, _, ng2 = h2g2.shape
-        # nb x nloc x 3 x axis
-        h2g2m = torch.split(h2g2, self.axis_dim, dim=-1)[0]
-        # nb x nloc x axis x ng2
-        g1_13 = torch.matmul(torch.transpose(h2g2m, -1, -2), h2g2) / (3.0**1)
-        # nb x nloc x (axisxng2)
-        g1_13 = g1_13.view(nb, nloc, self.axis_dim * ng2)
-        return g1_13
-
-    def _update_g1_grrg(
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/descriptor/repformer_layer.py:494]: do not mv the position of grrg in this pr.
you may want to reused the symmetry op in all dp descriptors. refact it in a future PR.

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

File: deepmd/pt/model/descriptor/repformer_layer.py (hunk 8 above)
Comment (iProzd) [reply deepmd/pt/model/descriptor/repformer_layer.py:140]: Both update_g1_grrg and update_g1_drrd used this op (now in symmetry_op), maybe a more general name needed.

File: deepmd/pt/model/descriptor/dpa2.py [hunk 10]
```diff
@@ -75,118 +91,171 @@ def __init__(
...
-            float
-        ] = None,  # work around the bad design in get_trainer and DpLoaderSet!
-        sel: Optional[
-            int
-        ] = None,  # work around the bad design in get_trainer and DpLoaderSet!
+        seed: Optional[int] = None,
+        resnet_dt: bool = False,
+        trainable_ln: bool = True,
+        ln_eps: Optional[float] = 1e-5,
+        type_one_side: bool = False,
```
[the remark this answers is not in the document]
Comment (iProzd) [reply deepmd/pt/model/descriptor/dpa2.py]: https://github.com/deepmodeling/deepmd-kit/pull/3758/commits/e9fe376e99f9b6f8dae08bbeaa22f3557101a3ad moves `resnet_dt` and `type_one_side` into repinit, `trainable_ln` and `ln_eps` into repformer. These four args are used by only one block.

File: deepmd/pt/model/descriptor/dpa2.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (iProzd) [reply deepmd/pt/model/descriptor/dpa2.py:100]: Maybe two problems need to figure out:
1. input script and argcheck surely can be seperated into sub blocks, but in class method, we may need such translation classes to assign **default values**:
```
class DescrptDPA2(NativeOP, BaseDescriptor):
    def __init__(
        self,
        repinit_params: dict = {},
        repformer_params: dict = {},
        descriptor_params: dict = {},
    ):
        self.repinit_params = RepInitParams(**repinit_params)
        self.repformer_params = RepFormerParams(**repformer_params)
        self.descriptor_params = DescriptorParams(**descriptor_params)
```
or is there any better way?

2. A breaking change for old input script, haven't figured out how to solve it yet. Any idea? @njzjz @wanghan-iapcm

[the remark this answers is not in the document]
Comment (iProzd) [reply deepmd/pt/model/descriptor/dpa2.py:100]: This will be fixed in next PR.

**Review by njzjz — APPROVED** [review]

**Review by wanghan-iapcm — APPROVED** [review]

**Pull request merged**